legend-engine-xt-javaPlatformBinding-pure/src/main/resources/core_java_platform_binding/legendJavaPlatformBinding/planConventions/test/pureDateLibraryTests.pure

```diff
@@ -79,33 +79,33 @@ meta::pure::executionPlan::platformBinding::legendJava::library::pureDate::tests
          ->assert('%s.equals(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.newPureDate(2015,4,9))')
-      ->addTest('Previous Day of week', {| %2015-04-13->previousDayOfWeek(DayOfWeek.Friday)}, 'org.finos.legend.engine.plan.dependencies.util.Library.previousDayOfWeek(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.parsePureDate("2015-04-13"), org.finos.legend.engine.plan.dependencies.domain.date.DayOfWeek.Friday)', $conventions->className(PureDate))
+      ->addTest('Previous Day of week', {| %2015-04-13->previousDayOfWeek(DayOfWeek.Friday)}, 'org.finos.legend.engine.plan.dependencies.util.Library.previousDayOfWeek(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.parsePureDate("2015-04-13"), org.finos.legend.engine.plan.dependencies.domain.date.DayOfWeek.getEnumFromPureName("Friday"))', $conventions->className(PureDate))
          ->assert('%s.equals(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.newPureDate(2015,4,10))')
-      ->addTest('Previous Day of week', {| %2015-04-13->previousDayOfWeek(DayOfWeek.Saturday)}, 'org.finos.legend.engine.plan.dependencies.util.Library.previousDayOfWeek(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.parsePureDate("2015-04-13"), org.finos.legend.engine.plan.dependencies.domain.date.DayOfWeek.Saturday)', $conventions->className(PureDate))
+      ->addTest('Previous Day of week', {| %2015-04-13->previousDayOfWeek(DayOfWeek.Saturday)}, 'org.finos.legend.engine.plan.dependencies.util.Library.previousDayOfWeek(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.parsePureDate("2015-04-13"), org.finos.legend.engine.plan.dependencies.domain.date.DayOfWeek.getEnumFromPureName("Saturday"))', $conventions->className(PureDate))
          ->assert('%s.equals(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.newPureDate(2015,4,11))')
-      ->addTest('Previous Day of week', {| %2015-04-13->previousDayOfWeek(DayOfWeek.Sunday)}, 'org.finos.legend.engine.plan.dependencies.util.Library.previousDayOfWeek(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.parsePureDate("2015-04-13"), org.finos.legend.engine.plan.dependencies.domain.date.DayOfWeek.Sunday)', $conventions->className(PureDate))
+      ->addTest('Previous Day of week', {| %2015-04-13->previousDayOfWeek(DayOfWeek.Sunday)}, 'org.finos.legend.engine.plan.dependencies.util.Library.previousDayOfWeek(org.finos.legend.engine.plan.dependencies.domain.date.PureDate.parsePureDate("2015-04-13"), org.finos.legend.engine.plan.dependencies.domain.date.DayOfWeek.getEnumFromPureName("Sunday"))', $conventions->className(PureDate))
```

Comment:
Out of curiosity - how this compare on efficiency/performance?  We went from a direct access to a call that requires a lookup on some  sort of collection.

Reply:
if the name contains special characters than it cannot be a java identifier, Thats why I am maintaining {enumConstant, pureName} inside enum.
so the actual name of enum will be different than the pureName(which may contain special characters). something like this
    ENUM0("Source Enum@$#_*[]{}()");  this type of generation happens only if one or more enums have special character
   ENUM1("Source Enum $$^$^^$");
   
   getEnumFromPureName() just uses a hashmap to look for an enumConstant for correspding pureName, Since it's O(1) operation I don't think it should impact much.

Reply:
Yeah we can directly access the hashMap rather than calling the function, I will change it, thanks for suggesting.

Reply:
Not sure I follow what you mean on the hashMap.  I meant, can we generate `EnumClass.ENUM_CONSTANT` as we were doing before?

Reply:
No we can't do this EnumClass.ENUM_CONSTANT  because we don't know what the  Java Enumconstant will be for corresponding pure EnumConstant, We have to look at the hashmap which i am generating.
here is the java enum that is being generated.

public enum SourceEnum
{
    ENUM0("Source Enum@$#_*[]{}()");

    private String pureName;
    private static final Map<String, SourceEnum> pureNameToEnumMap = initValues();

    SourceEnum(String pureName)
    {
        this.pureName = pureName;
    }

    public String getName()
    {
        return this.pureName;
    }

    public static SourceEnum getEnumFromPureName(String pureName)
    {
        if (pureName == null)
        {
            throw new NullPointerException("pureName of Enum cannot be a null");
        }
        else if (SourceEnum.pureNameToEnumMap.containsKey(pureName))
        {
            return SourceEnum.pureNameToEnumMap.get(pureName);
        }
        else
        {
            throw new IllegalArgumentException("Unmapped value: " + pureName);
        }
    }

    private static Map<String, SourceEnum> initValues()
    {
        Map<String, SourceEnum> values = new HashMap<String, SourceEnum>();
        values.put("Source Enum@$#_*[]{}()", SourceEnum.ENUM0);
        return values;
    }
}

Reply:
As discussed, we seem to be optimizing plan generation over plan execution.  Whenever possible, we should try to optimize plan execution and absorb any cost up-front during plan generation.  

If we create a common function, `pureEnumToJavaName(e: Enum[1])`, we can easily always reference either enum name during Java code generation, avoiding the need for an extra map (or at least the lookup on the map).  

I understand performance of this might not be ideal, but most enums are a handful of items, so linear lookup should not be as bad, and this will be done during generation.

Reply:
@rafaelbey At the time of de-serializing an enum property from a JSON object / serializing an enum property, names with special characters need to be used. I don't get how we can get rid of this check on names with special characters at execution time. Can you please explain a bit more?

Reply:
1. Discussed this issue of de-serialization with @rafaelbey, For deserialization we need to maintain a hashmap(pureName->javaEnum), But we can reduce the hashmap lookups for enum access, Because javaEnum name is derived from the index of the enum. Hence we can make such function pureEnumToJavaName(e: Enum[1]).
2. Also Maintaining a hashmap inside an enum for deserialisation should not create any performance issues because currenlty we use valueOf() function for deserialisation, which itself creates a hashmap(javaName->javaEnum).

Reply:
Exactly!  Deserializing will always take more time given the lookups to go from json to Java.  But in reverse, we have the ability to make the code more efficient by avoiding intermediaries, and go to the value we want to serialize.

Reply:
Yes, For serialization we are not doing any hashmap lookups, we have pureName variable inside the enumClass which is directly used for serialization.

Reply:
As discussed, Hashmap lookups should not create a significant impact on the execution. Also if we change the implementation then there will be a significant impact on the plan generation and it will also increase the code complexity. Hence we are moving with the current changes.